### DISCUSSION TITLE
version 4.0.0

### DISCUSSION THREAD

**Discussion by hvianna** [hvianna/audioMotion-analyzer#45] [Announcements]
### 🚨 BREAKING CHANGES:

+ Removed deprecated properties:
  + ~`energy`~ - use [`getEnergy()`](README.md#getenergy-preset-startfreq-endfreq-) instead
  + ~`isLedDisplay`~ - use [`isLedBars`](README.md#isledbars-boolean-read-only) instead
  + ~`peakEnergy`~ - use [`getEnergy('peak')`](README.md#getenergy-preset-startfreq-endfreq-) instead
  + ~`showLeds`~ - use [`ledBars`](README.md#ledbars-boolean) instead

+ [`getBars()`](README.md#getbars) now includes an additional `freq` property for each element, representing the center frequency of the band.
In addition, the range of each band has been adjusted so that frequencies that were previously at the lower edge (`freqLo`) are now located at the center of the band.

+ **`stereo` has been DEPRECATED** and will be removed in version 5 - use [`channelLayout`](README.md#channellayout-string) instead.

### Added:

+ [`ansiBands`](README.md#ansibands-boolean) - use ANSI/IEC preferred frequencies for the octave bands;
+ [`channelLayout`](README.md#channellayout-string) - adds option for dual channel combined spectrum, with both channel graphs overlaid;
+ [`frequencyScale`](README.md#frequencyscale-string) - adds linear and perceptual (Bark/Mel) frequency scales;
+ [`gradientLeft`](README.md#gradientleft-string) and [`gradientRight`](README.md#gradientright-string) - allow to select different gradients for each channel;
+ [`isBandsMode`](README.md#isbandsmode-boolean-read-only) read-only property;
+ [`linearAmplitude`](README.md#linearamplitude-boolean) - use linear values instead of dB for spectrum amplitudes;
+ [`linearBoost`](README.md#linearboost-number) - amplify low energy values when using linear amplitude;
+ [`noteLabels`](README.md#notelabels-boolean) - display musical note labels instead of frequency values on the X-axis;
+ [`weightingFilter`](README.md#weightingfilter-string) - select from five different weighting filters for improved spectrum visualization;
+ Two new built-in [gradients](README.md#gradient-string): _orangered_ and _steelblue_.

Thank you **@jonathan-annett** [(#28)](https://github.com/hvianna/audioMotion.js/issues/28) and **@TF3RDL** ([#30](https://github.com/hvianna/audioMotion-analyzer/issues/30) and [#38](https://github.com/hvianna/audioMotion-analyzer/issues/38)) for the suggestions that led to the implementation of several of these new features.

### Fixed:

+ LED peaks showing below zero level when `reflexRatio` > 0.

### Changed and improved:

+ Optimized generation of octave bands, with more accurate ranges and center frequencies;
+ Frequency labels on the X-axis now show different values depending on the value of [`ansiBands`](README.md#ansibands-boolean), to properly match the bands' center frequencies;
+ The [`onCanvasDraw`](README.md#oncanvasdraw-function) callback is now passed an additional object with timestamp and gradients information;
+ The _prism_ and _rainbow_ gradients were recreated using the beautiful [12-bit rainbow palette](https://iamkate.com/data/12-bit-rainbow/) by Kate Morley - If you need the old colors [check this post](https://github.com/hvianna/audioMotion-analyzer/discussions/44);
+ Custom gradients can now register a single color;
+ [`splitGradient`](README.md#splitgradient-boolean) now works for horizontal gradients in [`radial`](README.md#radial-boolean) visualization;
+ [`maxFreq`](README.md#maxfreq-number) is now capped to half the AudioContext's sample rate (Nyquist frequency);
+ Export additional types and interfaces in the TypeScript definition file;
+ Code optimizations, minor fixes and clean-up;
+ Revised and improved demos and documentation.

<hr /><em>This discussion was created from the release <a href='https://github.com/hvianna/audioMotion-analyzer/releases/tag/4.0.0'>version 4.0.0</a>.</em>

**Comment by evoyy** [comment]
Congratulations on the new release! I upgraded to v4 with no issues.

I tried the new option `channelLayout` set to `dual-combined`. I think it's an interesting idea, though it is not very visually effective since both channels blend together. I wonder if any of the canvas composite operations would help here. Maybe you tried that?

Also, I prefer the old rainbow gradient; it's more colorful and the new one seems a bit dull. Maybe that's just what I'm used to. Thanks for showing how to keep it, that was useful.

**Reply by hvianna** [reply]
Thanks @evoyy! 🙌 

Indeed, the `dual-combined` layout currently works best with `fillAlpha` < 1 (on Graph mode or using `outlineBars`) so you can see both channels.

I did some initial testing with an option to control the `globalCompositeOperation`, but it needs more work. It's on my to-do list for future releases though.

Thanks again for the feedback!

**Reply by evoyy** [reply]
`fillAlpha` does help a bit. I guess it also depends on the audio as well, the difference in L/R channels is often very subtle. A crazy idea would be to give the analyzer a z-plane and put one channel slightly behind the other...

I have added audioMotion-analyzer to [BAT.directory](https://bat.directory/), which is a directory/database of Brave verified content creators I am developing.
